File: test/features/sell/cart_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:voxon90/core/format/money_format.dart';
import 'package:voxon90/features/sell/domain/cart.dart';

const soda = {
  'id': 'p-soda',
  'name': 'Refresco',
  'priceCents': 5900,
  'unit': 'unit',
  'allowsFraction': false,
};
const rice = {
  'id': 'p-rice',
  'name': 'Arroz',
  'priceCents': 3500,
  'unit': 'pound',
  'allowsFraction': true,
};

void main() {
  test('agrega productos, suma cantidades y estima el importe', () {
    final cart = const Cart().add(soda).add(soda).add(rice, quantityMilli: 500);

    expect(cart.lines, hasLength(2));
    expect(cart.lines.first.quantityMilli, 2000);
    expect(cart.estimatedCents, 13550);
    expect(cart.toSaleLines(), [
      {'productId': 'p-soda', 'quantityMilli': 2000},
      {'productId': 'p-rice', 'quantityMilli': 500},
    ]);
  });

  test('respeta las unidades enteras y rechaza cantidades inválidas', () {
    expect(
      () => const Cart().add(soda, quantityMilli: 500),
      throwsA(
        isA<CartException>().having((e) => e.code, 'code', 'whole_units'),
      ),
    );
    final cart = const Cart().add(soda);
    expect(
      () => cart.setQuantity('p-soda', 1500),
      throwsA(isA<CartException>()),
    );
    expect(
      () => cart.add(rice, quantityMilli: 0),
      throwsA(
        isA<CartException>().having((e) => e.code, 'code', 'invalid_quantity'),
      ),
    );
  });

  test('poner cero quita la línea y el carrito original no cambia', () {
    final original = const Cart().add(soda).add(rice, quantityMilli: 2000);
    final updated = original.setQuantity('p-soda', 0);

    expect(updated.lines.map((line) => line.productId), ['p-rice']);
    expect(original.lines, hasLength(2));
    expect(updated.remove('p-rice').isEmpty, isTrue);
  });

  test('formato de dinero y cantidades', () {
    expect(formatMoney(125075), 'RD\$1,250.75');
    expect(formatMoney(5), 'RD\$0.05');
    expect(formatMoney(-350), '-RD\$3.50');
    expect(formatMoney(123456789), 'RD\$1,234,567.89');
    expect(parseMoney('1,250.75'), 125075);
    expect(parseMoney('abc'), isNull);
    expect(formatQuantity(500), '0.5');
    expect(parseQuantity('0,5'), 500);
    expect(parseQuantity('-1'), isNull);
    expect(unitLabel('pound'), 'lb');
  });
}

File: backend/web/tests/cart.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';

import { Cart } from '../js/cart.js';

const soda = { id: 'p-soda', name: 'Refresco', priceCents: 5900, unit: 'unit', allowsFraction: false };
const rice = { id: 'p-rice', name: 'Arroz', priceCents: 3500, unit: 'pound', allowsFraction: true };

test('agrega productos, suma cantidades y estima el bruto', () => {
  const cart = new Cart();
  cart.add(soda);
  cart.add(soda);
  cart.add(rice, 500);

  assert.equal(cart.lines.length, 2);
  assert.equal(cart.lines[0].quantityMilli, 2000);
  assert.equal(cart.estimatedGrossCents(), 13550);
  assert.deepEqual(cart.toSaleLines(), [
    { productId: 'p-soda', quantityMilli: 2000 },
    { productId: 'p-rice', quantityMilli: 500 },
  ]);
});

test('respeta los productos que se venden por unidades enteras', () => {
  const cart = new Cart();
  assert.throws(() => cart.add(soda, 500), /unidades enteras/);
  cart.add(soda);
  assert.throws(() => cart.setQuantity('p-soda', 1500), /unidades enteras/);
  assert.throws(() => cart.add(rice, 0), /mayor que cero/);
});

test('cambiar a cero quita la línea y vaciar deja el carrito limpio', () => {
  const cart = new Cart();
  cart.add(soda);
  cart.add(rice, 2000);
  cart.setQuantity('p-soda', 0);
  assert.deepEqual(
    cart.lines.map((line) => line.productId),
    ['p-rice'],
  );

  // Las líneas devueltas son copias: modificarlas no cambia el carrito.
  cart.lines[0].quantityMilli = 1;
  assert.equal(cart.lines[0].quantityMilli, 2000);

  cart.clear();
  assert.equal(cart.isEmpty, true);
});
